commons/zenoh-buffers/src/zslice.rs

```diff
@@ -30,17 +31,59 @@ use core::{
 /*************************************/
 pub trait ZSliceBuffer: Send + Sync + fmt::Debug {
     fn as_slice(&self) -> &[u8];
+
+    #[cfg(feature = "shared-memory")]
+    /// # Safety
+    /// Get mutable access to underlying data. This is safe if
+    /// there is no sharing of underlying data.
+    unsafe fn as_mut_slice_unchecked(&mut self) -> &mut [u8];
+    #[cfg(feature = "shared-memory")]
+    fn as_mut_slice(&mut self) -> Option<&mut [u8]>;
```

Comment:
This feels very weird, we are changing the function return type depending on the feature. What's the real need for that?

Reply:
The opposite for Option is:

```
unsafe fn as_mut_slice_unchecked(&mut self) -> &mut [u8];
#[cfg(feature = "shared-memory")]
fn is_shared(&self) -> bool;
```

Need is in having both efficient and fail-safe trait interface:
- for SHM builds `as_mut_slice` is failable, as the underlying data might have additional indirection, that's why I use Option as a return type. This Option is a good way to do the following opts:
- - perform both "check" and "get" in one `dyn trait` function call
- - force user to "check" before "get"
- for non-SHM builds `as_mut_slice` is not failing, so I remove Option and all corresponding overhead.

The interface came out a little complicated, but I treat it as performant and safe to use.